src/System.Text.Json/src/System/Text/Json/Node/JsonObject.cs

```diff
@@ -0,0 +1,465 @@
+            : this(duplicatePropertyNameHandling)
+            => AddRange(jsonProperties);
+
+        /// <summary>
+        ///   Gets or sets the value of the specified property.
+        /// </summary>
+        /// <param name="propertyName">The property name of the value to get or set.</param>
+        /// <exception cref="ArgumentNullException">
+        ///   Provided property name is null.
+        /// </exception>
+        public JsonNode this[string propertyName]
+        {
+            get => propertyName != null ? GetPropertyValue(propertyName) : throw new ArgumentNullException(nameof(propertyName));
+            set => _dictionary[propertyName] = value ?? throw new ArgumentNullException(nameof(propertyName));
+        }
+
+        /// <summary>
+        ///   Returns an enumerator that iterates through the JSON object properties.
+        /// </summary>
+        /// <returns>An enumerator structure for the JSON object.</returns>
+        /// <exception cref="ArgumentException">
+        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public IEnumerator<KeyValuePair<string, JsonNode>> GetEnumerator() => _dictionary.GetEnumerator();
+
+        /// <summary>
+        ///   Adds the specified property to the JSON object.
+        /// </summary>
+        /// <param name="jsonProperty">The property to add.</param>
+        /// <exception cref="ArgumentException">
+        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        public void Add(KeyValuePair<string, JsonNode> jsonProperty) => Add(jsonProperty.Key, jsonProperty.Value);
+
+        /// <summary>
+        ///   Adds the specified <see cref="JsonNode"/> property to the JSON object.
+        /// </summary>
+        /// <param name="propertyName">Name of the property to add.</param>
+        /// <param name="propertyValue">Value of the property to add.</param>
+        /// <exception cref="ArgumentNullException">
+        ///   Provided property name is null.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   Property name to set already exists if handling duplicates is set to <see cref="DuplicatePropertyNameHandling.Error"/>.
+        /// </exception>
+        /// <remarks>
+        ///   Null property value is allowed and represents a null JSON node.
+        /// </remarks>
+        public void Add(string propertyName, JsonNode propertyValue)
+        {
+            if(propertyName == null)
+            {
+                throw new ArgumentNullException(nameof(propertyName));
+            }
+
+            if (_dictionary.ContainsKey(propertyName))
+            {
+                switch (_duplicatePropertyNameHandling)
```

Comment:
@ahsonkhan do you think it might be worth it to also add a case here for the `Update` scenario under a #if DEBUG that logs that we are performing an update on the property for debugging purposes?

Reply:
When you say "logs" what do you mean? To where?

Do you mean something like the following?
```C#
                switch (_duplicatePropertyNameHandling)
                {
                    case DuplicatePropertyNameHandling.Ignore:
                        return;
                    case DuplicatePropertyNameHandling.Error:
                        throw new ArgumentException(SR.JsonObjectDuplicateKey);
                }
                Debug.Assert(_duplicatePropertyNameHandling == DuplicatePropertyNameHandling.Replace);
```

Reply:
I agree with Ahson that I'm not sure what Jose's question is... but I also agree with Ahson that we want a Debug.Assert that we're getting the third of three enum values (makes it easier to find where we missed adding support for the fourth one).